ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt

```diff
@@ -199,6 +200,12 @@ internal class Endpoint(
         connections.incrementAndGet()
 
         try {
+            if (!Platform.isCleartextTrafficPermitted(host)) {
+                throw UnknownServiceException(
+                    "CLEARTEXT communication to $host not permitted by network security policy"
+                )
+            }
```

Comment:
Not very sure if we should put this check here.

Reply:
@yschimke Can you take a look?

Reply:
I don't know enough about CIO to comment, is this HTTP only here? Or could it be HTTPS also?

Probably worth adding tests to validate the behaviour for both cases.

Some extra context for OkHttp.  We position it 

- After application interceptors have a chance to modify the request, in case they have an AlwaysSecureInterceptor configured.
- Before we connect via various options - proxies, or IPv4/IPv6 fallback, which might involve connecting to an IP instead of a host. So we avoid working around the platform restrictions.